### PR TITLE
SCPI base class

### LINKED ISSUE #804: SCPIbaseMixin class instead of includeSCPI
Ideas regarding instruments with common controls include Mixin classes which contain controls.

The first element are the base SCPI commands (right now included in `Instrument`)

I propose the following transition path:

1. Create a Mixin class and write the proper documentation (right now, the includeSCPI is not in the documentation!). I propose `SCPIbaseMixin` as class name, as it contains the SCPI base commands.
2. Deprecate `includeSCPI`
3. Add the SCPIbaseMixin to all instruments with `includeSCPI=True`, add `SCPIbaseUnknownMixin` to all instruments without an explicit definition of `includeSCPI`. The latter class is a subclass of the former one and raises a deprecation warning, such that we get information, whether the instrument is SCPI conform (changing to the SCPIbaseMixin) or not (removing the Mixin).

During step 3, we can tackle #658 (already closed) as well, as both require touching the init of all instruments.

### PR DESCRIPTION
**Pull request by BenediktBurger**
This PR adds a SCPI base class to supplant `includeSCPI=True`.
- Closes #804
- Closes #587 
- See #825 

With this transition I propose to remove SCPI attribute as well (at the end).

ToDos
- [x] Implement a base SCPI instrument
- [x] Ensure correct working
- [x] All instruments in the codebase with (implicit or explicit) `includeSCPI=True` should inherit one of the mixins.
- [x] All instruments with a custom `check_errors` should transition to use `fetch_next_error`, if applicable. Done in #1024 

Transition path:
1. step
- `includeSCPI=True` (whether explicit or implicit as the default value) is deprecated and raises a warning, but continues to work as expected.
- All instruments in the codebase, which have `includeSCPI=True` inherit from the SCPImixin and do not define includeSCPI anymore.
- All instruments in the codebase without a `includeSCPI` parameter inherit from SCPIunknownMixin until the status is defined. That way they continue to work as expected.
- Instruments with `includeSCPI=False` stay as they are (no warning raised).
2. step
- `Instrument` will not offer SCPI methods anymore, neither the SCPI variable
- `includeSCPI` (any value) is deprecated (raising warnings), but won't do anything (i.e. will be effectively `includeSCPI=False`)
3. step
- removal of includeSCPI from the signature.
- Removal of the SCPI variable

### CODE REVIEW

File: pymeasure/instruments/generic_types.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        self.write("*RST")
+
+    def check_errors(self):
+        """ Read all errors from the instrument.
+
+        :return: list of error entries
+        """
+        errors = []
+        while True:
+            err = self.values("SYST:ERR?")
```
[the remark this answers is not in the document]
Comment (bmoneke) [reply pymeasure/instruments/generic_types.py]: I included a "fetch_next_error" method according to the discussion there.

[the remark this answers is not in the document]
Comment (msmttchr) [reply pymeasure/instruments/generic_types.py]: Two quick questions:
1) Is there a specific reason why next error is a method and not a property, following style in this file ?
2) general question: normally for scpi commands that read/write/ask from/to instrument, we use `Instrument`  property factory. Is there a reason why we don't use them in this file? (I have the feeling there is, but I am unable to focus it)

[the remark this answers is not in the document]
Comment (bmoneke) [reply pymeasure/instruments/generic_types.py]: I oriented myself at the consensus of the other PR, which had a method.

Looking at the discussion again, I notice that the important result was, to make `next_error` public. Regarding the topic of method or property you talked about a property, which was not implemented.

I'll change it to a property

**Comment by mcdo0486** [comment]
I'd include updating driver documentation `.rst` files for all drivers to inherit `SCPImixin` so the SCPI properties are included on their instrument driver page.

**Comment by bilderbuchi** [comment]
I suspect if we do "cooperative multiple inheritance" the order of classes inherited probably should not matter, but I haven't confirmed this gut feeling.
This article is great (like everything by Raymond Hettinger) https://rhettinger.wordpress.com/2011/05/26/super-considered-super/ 
[another one I just found by google, haven't read](https://medium.com/swlh/cooperative-multiple-inheritance-paradigm-in-python-f048b7ecdb29))

**Comment by bmoneke** [comment]
As long as the init (if any) is cooperative and there are no overridden methods, the order does not matter.

In this case (Scpi and instrument) , however, the order does matter, as Scpi overrides the check errors method.

**Comment by bilderbuchi** [comment]
Ah, of course! I was purely thinking of the inits 😆

**Comment by BenediktBurger** [comment]
From my point of view, this `SCPIMixin` is ready to use.

The changes to all the instruments happens in another PR.

Should we write the new `adding_instruments` in this PR or in a separate one?

**Comment by bilderbuchi** [comment]
> Should we write the new `adding_instruments`in this PR or in a separate one?

not sure what you mean exactly with `adding_instruments`, but smaller PRs are generally preferable (for separable features). So, I'd make a separate one. :-)

**Comment by BenediktBurger** [comment]
> Should we write the new adding_instruments in this PR or in a separate one?

The documentation regarding adding new instruments has to be adjusted to the new Mixin.

It is already written as a task for the PR changing all instruments (#1019), so it won't be forgotten.

File: pymeasure/instruments/generic_types.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    @property
+    def options(self):
+        """ Get the device options installed. """
+        return self.ask("*OPT?").strip()
+
+    @property
+    def id(self):
+        """ Get the identification of the instrument. """
+        return self.ask("*IDN?").strip()
+
```
Comment (msmttchr) [inline pymeasure/instruments/generic_types.py:66]: @BenediktBurger, I think I have already asked this (sorry for the repetition), but why we are using explicit `@property` and not property factory `Instrument.measurement` ? (same comment apply to all `@property` above). Is it to keep minimal changes/risks ?

Comment (BenediktBurger) [reply pymeasure/instruments/generic_types.py:66]: I just moved the instrument`s methods here.

You're right, we can change them to properties.

Comment (BenediktBurger) [reply pymeasure/instruments/generic_types.py]: I pushed changes.

**Review by msmttchr — APPROVED** [review]
Ok for me and really thank you for all your efforts.

**Pull request merged**